### PR TITLE
Benchmark executable and script

### LINKED ISSUE #1439: Create scaffolding to make it easy to write ppx_bench benchmarks

### PR DESCRIPTION
**Pull request by psteckler**
Provide a convenient way to run benchmarks using the `ppx_bench` syntax extension. This is meant as a step toward resolving #2663. Closes #1439.

A benchmark can be added to any library using the `let%bench` or `let%bench_fun` syntaxes, and grouped with `let%bench_module`. The executable is built from `app/benchmarks/main.ml`. If a new library is benchmarked, you'd add that to the libraries in `app/benchmarkes/dune`, and to the list in `main.ml`. You build the executable with `make benchmarks`.

The script `scripts/benchmarks.sh` runs the executable with some flags. You can add other flags as needed. Running the script with `-help` shows what flags are available.

There are benchmarks for Pedersen hashing in this PR. Here's the output with no additional flags given:
```
Running inline tests in library "snark_params"
Estimated testing time 1m20s (8 benchmarks x 10s). Change using '-quota'.
┌─────────────────────────────────────────────────────────────────────────────────────────┬────────────┬──────────┬────────────┬────────────┬────────────┐
│ Name                                                                                    │   Time/Run │  mWd/Run │   mjWd/Run │   Prom/Run │ Percentage │
├─────────────────────────────────────────────────────────────────────────────────────────┼────────────┼──────────┼────────────┼────────────┼────────────┤
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash one triple unchunked              │     2.47us │   1.95kw │      1.71w │      1.88w │      0.16% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash one triple chunked                │     2.51us │   2.08kw │      9.54w │      9.54w │      0.16% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash small number of triples unchunked │    38.10us │   8.95kw │    317.95w │    317.95w │      2.47% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash small number of triples chunked   │    16.00us │   5.98kw │     83.57w │     83.57w │      1.04% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash large number of triples unchunked │   386.52us │  80.06kw │  3_535.80w │  3_535.80w │     25.02% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash large number of triples chunked   │   138.33us │  42.69kw │    887.13w │    887.13w │      8.95% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash huge number of triples unchunked  │ 1_544.97us │ 318.36kw │ 16_355.04w │ 15_354.04w │    100.00% │
│ [lib/snark_params/snark_params.ml:Tick.Pedersen] hash huge number of triples chunked    │   579.03us │ 159.69kw │  5_707.72w │  4_706.72w │     37.48% │
└─────────────────────────────────────────────────────────────────────────────────────────┴────────────┴──────────┴────────────┴────────────┴────────────┘
```

### CODE REVIEW

**Comment by cmr** [comment]
It's been a long while since I looked at the benchmark runner, but does it know how to output machine-readable formats like CSV or something?

File: src/lib/snark_params/snark_params.ml [hunk 1]
```diff
@@ -472,8 +472,20 @@ module Tick = struct
...
         let result, unchunked_result = run_updates fold in
         assert (equal_states result unchunked_result)
+
+      let hash_unchunked n =
+        ignore (Lazy.force chunk_table) ;
+        let fold = gen_fold n in
+        State.update_fold_unchunked initial_state fold
+
+      let hash_chunked n =
+        ignore (Lazy.force chunk_table) ;
```
Comment (enolan) [inline src/lib/snark_params/snark_params.ml]: It'd be better to use `bench_fun` as describe [here](https://github.com/janestreet/ppx_bench/blob/f267f868b6e666899ab4abf771e6db8842c8cb5a/example/ppx_bench_sample.ml#L29) to set up the chunk table.

Comment (psteckler) [reply src/lib/snark_params/snark_params.ml]: OK, done.

**Comment by psteckler** [comment]
> Can we turn on the error estimate display by default?

I don't want to put that in the script, because I'm not sure there's a way to disable it once it's there.

You can pass the needed flags, if desired, maybe from another script that contains those flags.

File: src/lib/snark_params/snark_params.ml [hunk 2]
```diff
@@ -487,6 +497,33 @@ module Tick = struct
...
     let%test_unit "hash large number of chunks plus 2" =
       For_tests.run_hash_test ((Chunked_triples.Chunk.size * 250) + 2)
+
+    (* benchmark unchunked, chunked hashes *)
+
+    let%bench "hash one triple unchunked" = For_tests.hash_unchunked 1
+
+    let%bench_fun "hash one triple chunked" =
+      (* make sure chunk table deserialized *)
+      ignore (Lazy.force chunk_table) ;
```
Comment (enolan) [inline src/lib/snark_params/snark_params.ml]: Everything that uses the chunk table should use the `bench_fun` thing. If you can run only one benchmark, or we change the ordering later we could get bogus results.

Comment (psteckler) [reply src/lib/snark_params/snark_params.ml]: OK

**Pull request merged**